### ISSUE TITLE
Use v4l2 ioctls directly

### ISSUE THREAD

**Issue by DemiMarie** [elliotkillick/qubes-video-companion#10]
`v4l2-ctl`’s output format appears to be meant for human use, rather than programmatic use.  A better approach would be to use the v4l2 ioctls directly.

**Comment by elliotkillick** [comment]
I totally agree and knew this when I was working with `v4l2-ctl`, however, at the time I was really just looking for a quick solution without getting into C ioctls.

PRs are welcome!

**Comment by elliotkillick** [comment]
Actually, I decided I'm going to implement this now instead of having to do something in Python only to replace it with a better solution.

I already have working C code for getting the webcam formats, now I just need to get the dimensions and FPS of each then implement a function to get the best format.

**Comment by DemiMarie** [comment]
> Actually, I decided I'm going to implement this now instead of having to do something in Python only to replace it with a better solution.

:+1:

> I already have working C code for getting the webcam formats, now I just need to get the dimensions and FPS of each then implement a function to get the best format.

Is the plan to then construct an appropriate GStreamer pipeline accordingly?

**Comment by elliotkillick** [comment]
> Is the plan to then construct an appropriate GStreamer pipeline accordingly?
Yes, I'm currently working on this C program.

**Comment by DemiMarie** [comment]
> > Is the plan to then construct an appropriate GStreamer pipeline accordingly?
> 
> Yes, I'm currently working on this C program.

Does this mean that `qvc.Webcam` will be written entirely in C, including the UI?  Or do you plan on using Python too?

**Comment by elliotkillick** [comment]
Currently the plan for the C program is to implement the webcam best format setter program as well as something to call to replace instances of `v4l2-ctl` in the Python scripts. No plans to replace the Python scripts.
